File: convert_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import re
import pandas as pd
import yaml

# ---- Конфигурация базовых правил ----
# Здесь задаются политики, которые должны присутствовать всегда.
BASE_SRCKEYWORD = 'все поды в namespace'
DEFAULT_RULES = {
    'd8-monitoring/app:prometheus': {
        'direction': 'ingress',
        'port': '15020',
        'protocol': 'TCP',
        'labels': {
            'io.kubernetes.pod.namespace': 'd8-monitoring',
            'app.kubernetes.io/name': 'prometheus'
        }
    },
    'kube-system/node-local-dns': {
        'direction': 'egress',
        'port': '53',
        'protocol': 'UDP',
        'labels': {
            'io.kubernetes.pod.namespace': 'kube-system',
            'k8s-app': 'node-local-dns'
        },
        'rules': {  # Дополнительные правила DNS должны идти внутри блока toPorts
            'dns': [
                {'matchPattern': '*'}
            ]
        }
    },
    'd8-istio/istiod': {
        'direction': 'egress',
        'port': '15012',
        'protocol': 'TCP',
        'labels': {
            'io.kubernetes.pod.namespace': 'd8-istio',
            'app': 'istiod'
        }
    }
}

# ---- Утилиты для разбора и сборки правил ----

def split_labels(raw: str) -> dict:
    """
    Преобразует строку меток вида 'k1:v1,k2:v2' или с переносами строк
    в словарь {k1: v1, k2: v2}.
    """
    parts = re.split(r'[,\n]+', raw)
    labels = {}
    for part in parts:
        part = part.strip()
        if not part:
            continue
        if ':' in part:
            key, val = part.split(':', 1)
            labels[key.strip()] = val.strip()
    return labels


def detect_type(value: str) -> (str, str):
    """
    Определяет тип строки: CIDR, FQDN или label.
    Возвращает кортеж (тип, значение).
    """
    v = value.strip()
    if re.match(r'^\d+\.\d+\.\d+\.\d+/\d+$', v):
        return 'cidr', v
    if '.' in v and ':' not in v:
        return 'fqdn', v.lower()
    return 'label', v


def make_ingress_rule(entry: dict) -> dict:
    """
    Формирует один ingress блок для CiliumNetworkPolicy.
    Вложение 'rules' происходит внутри объекта toPorts.
    """
    rule = {}

    # FROM
    if 'labels' in entry:
        rule['fromEndpoints'] = [{'matchLabels': entry['labels']}]
    else:
        dtype, val = detect_type(entry['source'])
        if dtype == 'label':
            rule['fromEndpoints'] = [{'matchLabels': split_labels(val)}]
        elif dtype == 'fqdn':
            rule['fromFQDNs'] = [{'matchName': val}]
        else:
            rule['fromCIDRSet'] = [{'cidr': val}]

    # TO PORTS + вложенные правила (если есть)
    port_block = {
        'ports': [
            {
                'port': str(entry['port']),
                **({'protocol': entry['protocol'].upper()} if entry['protocol'].upper() != 'ANY' else {})
            }
        ]
    }
    if 'rules' in entry:
        # Вложение блока rules внутри toPorts
        port_block['rules'] = entry['rules']
    rule['toPorts'] = [port_block]

    return rule


def make_egress_rule(entry: dict) -> dict:
    """
    Формирует один egress блок для CiliumNetworkPolicy.
    Аналогично ingress.
    """
    rule = {}

    # TO
    if 'labels' in entry:
        rule['toEndpoints'] = [{'matchLabels': entry['labels']}]
    else:
        dtype, val = detect_type(entry['destination'])
        if dtype == 'label':
            rule['toEndpoints'] = [{'matchLabels': split_labels(val)}]
        elif dtype == 'fqdn':
            rule['toFQDNs'] = [{'matchName': val}]
        else:
            rule['toCIDRSet'] = [{'cidr': val}]

    # TO PORTS + вложенные правила
    port_block = {
        'ports': [
            {
                'port': str(entry['port']),
                **({'protocol': entry['protocol'].upper()} if entry['protocol'].upper() != 'ANY' else {})
            }
        ]
    }
    if 'rules' in entry:
        port_block['rules'] = entry['rules']
    rule['toPorts'] = [port_block]

    return rule


def parse_excel(path: str) -> (str, dict):
    """
    Читает Excel-файл без заголовков.
    Namespace во второй строке (колонка А).
    Группирует правила по ключу из колонки C.
    """
    df = pd.read_excel(path, header=None, dtype=str).fillna('')
    raw_ns = df.iat[1, 0].strip()
    if ':' not in raw_ns:
        sys.exit("Ошибка: не найден 'namespace:' во второй строке")
    namespace = raw_ns.split(':', 1)[1].strip()

    groups = {}
    for i in range(2, len(df)):
        row = df.iloc[i]
        if not str(row[0]).strip():
            continue
        entry = {
            'protocol': row[1].strip(),
            'source': row[2].strip(),
            'direction': row[3].strip().lower(),
            'destination': row[4].strip(),
            'port': row[5].strip()
        }
        key = row[2].strip() or BASE_SRCKEYWORD
        groups.setdefault(key, []).append(entry)

    return namespace, groups


def build_policy(namespace: str, key: str, rules: list) -> dict:
    """
    Собирает CiliumNetworkPolicy dict для одной группы.
    """
    labels = split_labels(key) if key != BASE_SRCKEYWORD else {}
    suffix = next(iter(labels.values())) if labels else 'default'
    policy = {
        'apiVersion': 'cilium.io/v2',
        'kind': 'CiliumNetworkPolicy',
        'metadata': {'name': f"{namespace}-{suffix}", 'namespace': namespace},
        'spec': {'endpointSelector': labels and {'matchLabels': labels} or {}}
    }

    ingress_list, egress_list = [], []
    for entry in rules:
        if entry['direction'] == 'ingress':
            ingress_list.append(make_ingress_rule(entry))
        else:
            egress_list.append(make_egress_rule(entry))

    if ingress_list:
        policy['spec']['ingress'] = ingress_list
    if egress_list:
        policy['spec']['egress'] = egress_list

    return policy


def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <input.xlsx>")
        sys.exit(1)

    # Парсим Excel
    namespace, groups = parse_excel(sys.argv[1])

    # Лог создания файла
    output = f"{namespace}.yaml"
    print(f"Файл создан: {output}")

    # Проверка базовых правил в Excel
    found = [name for name, cfg in DEFAULT_RULES.items()
             if any(r['destination'] == name and r['direction'] == cfg['direction']
                    for grp in groups.values() for r in grp)]
    if len(found) == len(DEFAULT_RULES):
        print("Базовые правила обнаружены и добавлены в конфигурацию.")
    else:
        print("Не обнаружены базовые правила, добавлены в конфигурацию.")
        # Заменяем группу default только на базовые правила
        groups[BASE_SRCKEYWORD] = []
        for name, cfg in DEFAULT_RULES.items():
            entry = {
                'protocol': cfg['protocol'],
                'source': BASE_SRCKEYWORD,
                'direction': cfg['direction'],
                'destination': name,
                'port': cfg['port'],
                'labels': cfg['labels']
            }
            if 'rules' in cfg:
                entry['rules'] = cfg['rules']
            groups[BASE_SRCKEYWORD].append(entry)

    # Генерируем политики: default сначала
    order = [BASE_SRCKEYWORD] + [k for k in groups if k != BASE_SRCKEYWORD]
    policies = [build_policy(namespace, k, groups[k]) for k in order]

    # Сохраняем в YAML с явными отступами
    with open(output, 'w', encoding='utf-8') as f:
        yaml.dump_all(
            policies, f,
            sort_keys=False,
            allow_unicode=True,
            default_flow_style=False,  # блочный стиль
            indent=2  # два пробела для вложенности
        )

if __name__ == '__main__':
    main()

File: convert_test.js
#!/usr/bin/env node

/**
 * Скрипт конвертации Excel → CiliumNetworkPolicy YAML
 * Использует:
 *   npm install xlsx js-yaml
 *
 * Особенности:
 * - Чтение дефолтных политик из default.yaml
 * - Проверка расширения .xlsx и существования файла
 * - Валидация наличия номеров правил в колонке A
 * - Проверка отсутствия кириллицы, поиск дубликатов
 * - Поддержка namespace/label:val для селекторов
 * - Обработка множественных портов и IP/CIDR-списков
 * - Подтверждение перезаписи выходного YAML
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const XLSX = require('xlsx');
const yaml = require('js-yaml');

// ---- Загрузка default.yaml и сбор базовых правил ----
const defaultYamlPath = path.join(__dirname, 'default.yaml');
if (!fs.existsSync(defaultYamlPath)) throw new Error('❌ Не найден default.yaml');
const defaultDocs = yaml.loadAll(fs.readFileSync(defaultYamlPath, 'utf8'));
const DEFAULT_RULES = defaultDocs.flatMap(doc => {
  const regs = [];
  (doc.spec.ingress || []).forEach(rule => {
    const p = rule.toPorts[0].ports[0];
    regs.push({ direction: 'ingress', port: String(p.port), protocol: (p.protocol||'ANY').toUpperCase() });
  });
  (doc.spec.egress || []).forEach(rule => {
    const p = rule.toPorts[0].ports[0];
    regs.push({ direction: 'egress', port: String(p.port), protocol: (p.protocol||'ANY').toUpperCase() });
  });
  return regs;
});

// ---- Утилиты ----
const splitLabels = raw => raw
  .split(/[\n,]+/)             // по запятым или переносам строки
  .map(s => s.trim())
  .filter(s => s.includes(':'))
  .reduce((o, p) => {
    const [k, ...v] = p.split(':');
    o[k.trim()] = v.join(':').trim();
    return o;
  }, {});

const detectType = v0 => {
  const v = (v0||'').trim();
  if (/^\d+\.\d+\.\d+\.\d+\/\d+$/.test(v)) return ['cidr', v];
  if (/^\d+\.\d+\.\d+\.\d+$/.test(v)) return ['cidr', v + '/32'];
  if (v.includes('.') && !v.includes(':')) return ['fqdn', v.toLowerCase()];
  return ['label', v];
};

/** Формирует правило ingress/egress */
const makeRule = e => {
  const { direction, sourceLabel, destLabel, port, protocol, rules } = e;
  const r = {};
  const prefix = direction === 'ingress' ? 'from' : 'to';

  // Спец-кейс namespace/label:val
  let labels;
  const slashField = direction === 'ingress' ? sourceLabel : destLabel;
  if (slashField && slashField.includes('/')) {
    const [nsPart, labPart] = slashField.split('/');
    labels = { 'io.kubernetes.pod.namespace': nsPart.trim(), ...splitLabels(labPart) };
  }

  // Селектор
  if (labels) {
    r[`${prefix}Endpoints`] = [{ matchLabels: labels }];
  } else {
    const keyField = direction === 'ingress' ? sourceLabel : destLabel;
    const keyList = String(keyField || '')
      .split(/[\n,]+/)    // разделяем по запятым и переносам строки
      .map(s => s.trim())
      .filter(Boolean);

    const [t0] = detectType(keyList[0] || '');
    if (t0 === 'label') {
      r[`${prefix}Endpoints`] = keyList.map(v => ({ matchLabels: splitLabels(v) }));
    } else if (t0 === 'fqdn') {
      r[`${prefix}FQDNs`] = keyList.map(v => ({ matchName: v }));
    } else {
      r[`${prefix}CIDRSet`] = keyList.map(v => ({ cidr: detectType(v)[1] }));
    }
  }

  // Порты → массив
  const portList = String(port)
    .split(/[\n,]+/)    // по запятым и переносам строки
    .map(s => s.trim())
    .filter(Boolean);
  const portsArray = portList.map(pv => {
    const po = { port: pv };
    if (protocol && protocol.toUpperCase() !== 'ANY') po.protocol = protocol.toUpperCase();
    return po;
  });
  const pb = { ports: portsArray };
  if (rules) pb.rules = rules;
  r.toPorts = [pb];

  return r;
};

// ---- Парсинг Excel и проверка номеров ----
function parseExcel(file) {
  const wb = XLSX.readFile(file);
  const rows = XLSX.utils.sheet_to_json(wb.Sheets[wb.SheetNames[0]], { header:1, blankrows:false });
  if (!rows[1] || !rows[1][0]) throw new Error('❌ Не найден namespace в строке 2');
  const nsRaw = String(rows[1][0]).trim();
  if (!nsRaw.includes(':')) throw new Error('❌ Неверный формат namespace');
  const namespace = nsRaw.split(':').slice(1).join(':').trim();

  const groups = {};
  const entries = [];
  const missingNums = [];

  for (let i = 2; i < rows.length; i++) {
    const row = rows[i] || [];
    const hasData = row.slice(1,6).some(c=>c && String(c).trim());
    if (hasData && !row[0]) missingNums.push(i+1);
    if (!hasData) continue;

    const protocol = String(row[1]||'').trim();
    const groupKey = String(row[2]||'').trim();
    const direction = String(row[3]||'').toLowerCase().trim();
    const other = String(row[4]||'').trim();
    const port = String(row[5]||'').trim();

    let sourceLabel = '', destLabel = '';
    if (direction === 'ingress') {
      sourceLabel = other;
      destLabel = groupKey;
    } else {
      sourceLabel = groupKey;
      destLabel = other;
    }

    const entry = { protocol, sourceLabel, destLabel, direction, port, rowNum: i+1 };
    entries.push(entry);
    (groups[groupKey] ||= []).push(entry);
  }

  return { namespace, groups, entries, missingNums };
}

// ---- Валидация ----
function validate(entries) {
  const bad = entries.filter(e=>/[А-Яа-яЁё]/.test(e.sourceLabel+e.destLabel+e.protocol+e.direction+e.port));
  if (bad.length) { console.error('❌ Кириллица на строках: '+bad.map(e=>e.rowNum).join(', ')); process.exit(1); }
  const seen = {}, dups = [];
  entries.forEach(e=>{
    const key = [e.direction, e.sourceLabel, e.destLabel, e.port, (e.protocol||'ANY').toUpperCase()].join('|');
    if (seen[key]) dups.push({a:seen[key],b:e.rowNum}); else seen[key]=e.rowNum;
  });
  if (dups.length) { console.warn('⚠️ Дубликаты:'); dups.forEach(d=>console.warn(`  строки ${d.a} и ${d.b}`)); }
}

// ---- Генерация политик ----
const buildPolicy = (ns, groupKey, ents) => {
  const labels = splitLabels(groupKey);
  const suffix = Object.values(labels)[0] || 'default';
  const name = `${ns}-${suffix}`;
  const spec = { endpointSelector: Object.keys(labels).length ? { matchLabels: labels } : {} };
  const ingress = [], egress = [];
  ents.forEach(e=>{
    const rule = makeRule(e);
    e.direction === 'ingress' ? ingress.push(rule) : egress.push(rule);
  });
  if (ingress.length) spec.ingress = ingress;
  if (egress.length) spec.egress = egress;
  return { apiVersion:'cilium.io/v2', kind:'CiliumNetworkPolicy', metadata:{name,namespace:ns}, spec };
};

// ---- Main ----
function main() {
  const file = process.argv[2];
  if (!file || !/\.xlsx$/i.test(file)) { console.error('❌ Укажите .xlsx файл'); process.exit(1); }
  if (!fs.existsSync(file)) { console.error(`❌ ${file} не найден`); process.exit(1); }

  const { namespace, groups, entries, missingNums } = parseExcel(file);
  if (missingNums.length) { console.error('⚠️ Нет номеров правил в строках: '+missingNums.join(', ')); console.error('❌ Отменено.'); process.exit(1); }

  validate(entries);

  const hasAll = DEFAULT_RULES.every(cfg =>
    entries.some(e => e.direction === cfg.direction && e.port === cfg.port && (cfg.protocol==='ANY'|| e.protocol.toUpperCase()===cfg.protocol))
  );
  console.log(hasAll ? '⚠️ Базовые есть' : '⚠️ Базовых нет');

  const defaultPolicies = defaultDocs.map(doc => ({
    ...doc,
    metadata: { name:`${namespace}-default`, namespace }
  }));

  const excelPolicies = Object.entries(groups).map(([k,ents]) => buildPolicy(namespace, k, ents));

  const allDocs = [...defaultPolicies, ...excelPolicies]
    .map(p => yaml.dump(p, { sortKeys:false, noRefs:true, indent:2, lineWidth:-1 }));

  const out = `${namespace}.yaml`;
  const write = () => { fs.writeFileSync(out, allDocs.join('---\n'),'utf8'); console.log(`✅ ${out} создан`); };

  if (fs.existsSync(out)) readline.createInterface({input:process.stdin,output:process.stdout})
    .question(`⚠️ ${out} уже существует. Перезаписать? (y/n): `, a=>{ a.toLowerCase().startsWith('y')?write():console.log('❌ Отменено.'); process.exit(0); });
  else write();
}

if (require.main===module) main();
